Decouple routing logic from address set

lutovich opened the pull request.

lutovich committed:
Decouple routing logic from address set

This commit introduces a separate load balancing strategy abstraction
responsible for address selection.

lutovich committed:
Simplified connection acquisition in cluster pool

Commit changes `TryAcquire` with out parameter to simple `Acquire` that
either returns a connection or null.

ali-ince commented on Neo4j.Driver/Neo4j.Driver/Internal/Routing/AddressSet.cs:
```diff
@@ -14,136 +14,105 @@
-                value = _items[_index++];
-            }
-            return true;
+            _items = new List<T>();
```
Should we also make this assignment a critical section?

lutovich replied:
Yes, good point. It is fine for `#Clear()` to assign because it is an atomic action, but it is not fine for it to not coordinate with other modification methods like `#Add()` or `#Remove()`.

ali-ince commented on Neo4j.Driver/Neo4j.Driver/Internal/Routing/IAddressSet.cs:
```diff
@@ -0,0 +1,65 @@
+        /// Not thread safe
+        /// </summary>
+        /// <returns></returns>
+        string ToString();
```
There's no need for a ```ToString()``` declaration as it is part of ```Object``` class that can be overrided.

ali-ince reviewed (approved):
Your changes looks ok. A few comments only.

lutovich committed:
Removed ToString from interface

lutovich committed:
Fixed locking in `AddressSet#Clear()`

Re-assignment of the internal list needs to be done while holding a lock
so that it is coordinated and isolated from other modifications like
`#Add()` or `#Remove()`.

lutovich commented:
@ali-ince comments should now be addresses. Thank you.
